### PR TITLE
NNVM Refactor

### PR DESCRIPTION
**Pull request by tqchen**
This is a PR that moves the symbolic layer to  https://github.com/dmlc/nnvm

This will enable new stage of mxnet's system optimization, by allowing more generic attributes and optimization passes that can be independent from each other. This refactor is backward compatible in API in a sense that all API is re-implemented with NNVM calls, without changing the interface.

Current changes:
- Move symbolic API to NNVM @tqchen
   - Most front-end C API are backward  compatible
   - Removed symbolic api in MXNet and relies on NNVM
- Known features removed limitations to be fixed
  - Bulk segment execution not yet added.
  - The gradient aggregation optimization hack by switching to addto is not yet added,
    can harm LSTM if it is constructed by unrolling the graph

Due to the major changes, the code will sit into a new branch first, before it get merged into master.

### CODE REVIEW

**Comment by tqchen** [comment]
I have tested the unittest cases and image classification CIFAR case.

File: Makefile [hunk 1]
```diff
@@ -140,6 +145,17 @@ else
 	SCALA_PKG_PROFILE := $(SCALA_PKG_PROFILE)-cpu
 endif
 
+# For quick compile test, used smaller subset
+ALLX_DEP = $(filter-out build/src/operator/%, $(ALL_DEP))
+ALLX_DEP+= build/src/operator/fully_connected.o
```
Comment (piiswrong) [inline Makefile]: remove hack?

File: Makefile [hunk 2]
```diff
@@ -181,13 +197,14 @@ build/plugin/%.o: plugin/%.cc
...
+lib/libmxnet.a: $(ALLX_DEP)
 	@mkdir -p $(@D)
 	ar crv $@ $(filter %.o, $?)
 
-lib/libmxnet.so: $(ALL_DEP)
+lib/libmxnet.so: $(ALLX_DEP)
 	@mkdir -p $(@D)
-	$(CXX) $(CFLAGS) -shared -o $@ $(filter %.o %.a, $^) $(LDFLAGS)
+	$(CXX) $(CFLAGS) -shared -o $@ $(filter %.o, $^) $(LDFLAGS) \
+	-Wl,--whole-archive $(filter %.a, $^) -Wl,--no-whole-archive
```
Comment (piiswrong) [inline Makefile]: this doesn't work on mac. use `-Wl,-force_load $(filter %.a, $^)` for mac

File: include/mxnet/executor.h [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+   * \param shared_exec input executor to share memory with.
+   * \return a new executor.
+   */
+  static Executor *Bind(nnvm::Symbol symbol,
+                        const Context& default_ctx,
+                        const std::map<std::string, Context>& group2ctx,
+                        const std::vector<NDArray> &in_args,
+                        const std::vector<NDArray> &arg_grad_store,
+                        const std::vector<OpReqType> &grad_req_type,
+                        const std::vector<NDArray> &aux_states,
```
Comment (piiswrong) [inline include/mxnet/executor.h]: should aux state still be here?
Removing it (keep compatibility by moving it to CAPI) would make interface more consistent with nnvm
however it breaks a few things.

File: include/mxnet/ndarray.h [hunk 4]
```diff
@@ -249,6 +249,22 @@ class NDArray {
     return ret;
   }
   /*!
+   * \brief Create a NDArray that shares memory with current one
+   *  The new array must have smaller memory size than the current array.
+   * \param shape new shape
+   * \param dtype The data type.
+   * \return NDArray in new shape and type.
+   */
+  inline NDArray AsArray(const TShape &shape, int dtype) const {
```
Comment (piiswrong) [inline include/mxnet/ndarray.h]: asarray seems like a misnomer.

File: include/mxnet/op_attr_types.h [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+/*!
+ * \brief Create a Layer style, forward/backward operator.
+ *  This is easy to write code that contains state.
+ *
+ *  This is not the only way to register an op execution function.
+ *  More simpler or specialized operator form can be registered
+ *
+ *  \note Register under "FCreateLayerOp"
+ */
+using FCreateLayerOp = std::function<
```
Comment (piiswrong) [inline include/mxnet/op_attr_types.h]: 'Layer' has never been used in mxnet. Rename back to operator?

File: src/nnvm/legacy_op_util.cc [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+    back_op.attr<nnvm::FMutateInputs>("FMutateInputs", OpBackMutateInputs);
+    back_op.attr<nnvm::FInplaceOption>("FInplaceOption", OpBackInplaceOption);
+    back_op.attr<FResourceRequest>(
+        "FResourceRequest", OpBackResourceRequest);
+    back_op.attr<bool>("TIsLayerOpBackward", true);
+  }
+}
+
+// no gradient operator
+NNVM_REGISTER_OP(_NoGradient)
```
Comment (piiswrong) [inline src/nnvm/legacy_op_util.cc]: what's this for?

File: src/executor/graph_executor.cc [hunk 7]
```diff
@@ -0,0 +1,599 @@
...
+    opnode.exec->op_ctx.is_train = is_train;
+    if (opnode.exec->exec_type() == Operator::kCrossDeviceCopy) {
+      CHECK_EQ(inode.inputs.size(), 1);
+      CHECK_EQ(opnode.exec->in_array.size(), 1);
+      CHECK_EQ(opnode.exec->out_array.size(), 1);
+      CopyFromTo(opnode.exec->in_array[0], &(opnode.exec->out_array[0]));
+    } else if (opnode.cached_opr != nullptr) {
+      Engine::Get()->Push(opnode.cached_opr, opnode.ctx);
+    } else {
+      LOG(FATAL) << "Not accessed";
```
Comment (piiswrong) [inline src/executor/graph_executor.cc]: You mean this line should never be reached?

File: src/executor/attach_op_execs_pass.cc [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+      mutate_index = fmutate_inputs[inode.source->op()](inode.source->attrs);
+    }
+    if (fcreate_layer_op.count(inode.source->op())) {
+      std::vector<TShape> ishape;
+      std::vector<int> itype;
+      for (const auto& e : inode.inputs) {
+        ishape.emplace_back(vshape[idx.entry_id(e)]);
+        itype.emplace_back(vdtype[idx.entry_id(e)]);
+      }
+      ret[i] = std::make_shared<ForwardOpExecutor>(
```
Comment (piiswrong) [inline src/executor/attach_op_execs_pass.cc]: Doesn't handle Ops without create layer (only has FCompute<gpu>)

File: src/executor/attach_op_execs_pass.cc [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+namespace mxnet {
+
+namespace op {
+const OperatorProperty* OpPropGetOpProperty(const NodeAttrs& attrs);
+}  // namespace op
+
+namespace exec {
+
+// forward executor
+class ForwardOpExecutor : public OpExecutor {
```
Comment (piiswrong) [inline src/executor/attach_op_execs_pass.cc]: Is this wrapper really necessary? There are too many layers of abstraction.
Can we try to squash some of them?

File: include/mxnet/tensor_blob.h [hunk 10]
```diff
@@ -20,512 +20,6 @@
 namespace mxnet {
 
 /*!
- * \brief A dynamic sized array data strcuture
```
Comment (piiswrong) [inline include/mxnet/tensor_blob.h]: Where is this moved to?

**Comment by piiswrong** [comment]
What were the reasons to move aux into input? Any reason beside allowing aux to be named by variable? Would it be better to handle it as a special case?

Looks like it's impossible to unify aux and input in the front end now due to legacy issues. Unifying it in the backend seems to only add confusions and complicated wrappers.

File: include/mxnet/op_attr_types.h (hunk 5 above)
Comment (tqchen) [inline include/mxnet/op_attr_types.h]: ForwardBackOp?

File: src/executor/graph_executor.cc (hunk 7 above)
Comment (tqchen) [inline src/executor/graph_executor.cc]: yes

File: src/executor/attach_op_execs_pass.cc (hunk 8 above)
Comment (tqchen) [inline src/executor/attach_op_execs_pass.cc]: FCompute has not yet been merged into this

File: src/executor/attach_op_execs_pass.cc (hunk 9 above)
Comment (tqchen) [inline src/executor/attach_op_execs_pass.cc]: It is helpful to unifies the Operator and FCompute interface. We can discuss if we want to move it to registry level

File: include/mxnet/tensor_blob.h (hunk 10 above)
Comment (tqchen) [inline include/mxnet/tensor_blob.h]: moved to here

**Comment by tqchen** [comment]
Moving aux to input simplifies the case for NNVM and make it general to handle most cases. It also allows  aux to be named by variable in a natural way. 

Handling aux in the front end  means differentiating mutable inputs without immutable ones.  While that is not a must thing to do, doing so also makes things clear in some case, and consistent with the original NNVM terminology.

As "aux = muttable input", the terminology part is clear and most ops don't need to have aux in the first place. Future ops don't need to rely on the translation, and interface is simplified(most ops don't need to mutate inputs)

We can discuss the terminology in frontend later but since one can be consistent with another, I feel it is fine.

**Comment by tqchen** [comment]
after discussion with @piiswrong We will merge this in as it already passed most tests.

**Comment by tqchen** [comment]
We will push followup updates to the nnvm branch.

@javelinjs We find that the scala test did not pass due to the fact that the newly added libnnvm.a is not linked, can you update the assembly script to add that?

**Comment by javelinjs** [comment]
@tqchen Sure, will update today.

**Pull request merged**